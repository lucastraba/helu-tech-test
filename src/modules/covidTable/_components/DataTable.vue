<template>
  <div
    class="overflow-auto"
  >
    <b-table
      v-if="(data.length || loading) && !errorMsg"
      sticky-header="100vh"
      striped
      show-empty
      foot-clone
      sort-direction="desc"
      no-footer-sorting
      :fields="fields"
      :items="data"
      :busy="loading"
      :filter="filterRegEx"
      :filter-included-fields="['countryName']"
      @filtered="setFilterResults"
    >
      <template #cell(cases)="column">
        {{ column.value | formatNumber }}
      </template>
      <template #cell(deaths)="column">
        {{ column.value | formatNumber }}
      </template>
      <template #cell(recoveries)="column">
        {{ column.value | formatNumber }}
      </template>
      <template #foot(countryName)="">
        <span class="text-dark">TOTALS</span>
      </template>
      <template #foot(cases)="">
        <span class="text-info">{{ totalCases | formatNumber }}</span>
      </template>
      <template #foot(deaths)="">
        <span class="text-danger">{{ totalDeaths | formatNumber }}</span>
      </template>
      <template #foot(recoveries)="">
        <span class="text-success">{{ totalRecoveries | formatNumber }}</span>
      </template>
      <template v-slot:table-busy>
        <div class="text-center text-danger my-2">
          <b-spinner class="align-middle" />
          <strong>Loading...</strong>
        </div>
      </template>
    </b-table>
    <div
      v-else
      class="text-danger"
    >
      {{ errorMsg }}
    </div>
  </div>
</template>

<script>
import TotalsMixin from '@/modules/covidTable/_components/TotalsMixin';

export default {
  name: 'DataTable',
  filters: {
    formatNumber (num) {
      return new Intl.NumberFormat().format(num);
    }
  },
  mixins: [TotalsMixin],
  props: {
    filter: {
      type: String,
      required: false,
      default: ''
    },
    data: {
      type: Array,
      required: false,
      default: () => []
    },
    errorMsg: {
      type: String,
      required: false,
      default: ''
    },
    loading: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  data () {
    return {
      fields: [{
        key: 'countryName',
        label: 'Country',
        sortable: true,
        sortDirection: 'asc'
      }, {
        key: 'cases',
        label: 'Cases',
        sortable: true
      }, {
        key: 'deaths',
        label: 'Deaths',
        sortable: true
      }, {
        key: 'recoveries',
        label: 'Recoveries',
        sortable: true
      }],
      filterResults: null
    };
  },
  computed: {
    totalCases () {
      return this.calculateTotals('cases', this.filterResults || this.data);
    },
    totalDeaths () {
      return this.calculateTotals('deaths', this.filterResults || this.data);
    },
    totalRecoveries () {
      return this.calculateTotals('recoveries', this.filterResults || this.data);
    },
    filterRegEx () {
      return new RegExp(`^${this.filter}`, 'i');
    }
  },
  watch: {
    filter (value) {
      if (value) {
        this.$nextTick(() => {
          this.$emit('suggestion', this.filterResults[0] && this.filterResults[0].countryName);
        });
      }
    }
  },
  methods: {
    setFilterResults (results) {
      this.filterResults = results;
    }
  }
};
</script>
