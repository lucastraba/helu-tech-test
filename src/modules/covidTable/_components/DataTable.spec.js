import { createLocalVue, shallowMount } from '@vue/test-utils';
import DataTable from './DataTable';

describe('DataTable Component', () => {
  const mock$emit = jest.fn();
  const localVue = createLocalVue();
  const factory = ({ propsData } = {}) => shallowMount(DataTable, {
    localVue,
    stubs: {
      bTable: true
    },
    mocks: {
      $emit: mock$emit
    },
    propsData: {
      ...propsData
    }
  });
  describe('conditional rendering', () => {
    it('should render the table provided there is data and there are no errors', () => {
      // Arrange.
      const wrapper = factory({
        propsData: {
          data: [1]
        }
      });
      // Act.
      // Assert.
      expect(wrapper.find('btable-stub').exists()).toBe(true);
      expect(wrapper.find('.error').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should render the table provided it is loading and there are no errors', () => {
      // Arrange.
      const wrapper = factory({
        propsData: {
          loading: true
        }
      });
      // Act.
      // Assert.
      expect(wrapper.find('btable-stub').exists()).toBe(true);
      expect(wrapper.find('.error').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should render the table provided it is loading and there are no errors', () => {
      // Arrange.
      const wrapper = factory({
        propsData: {
          errorMsg: 'foo'
        }
      });
      // Act.
      // Assert.
      expect(wrapper.find('btable-stub').exists()).toBe(false);
      expect(wrapper.find('.error').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('watch', () => {
    it('should emit a suggestion after a tick if it has value', async () => {
      // Arrange.
      const wrapper = factory();
      // Act.
      wrapper.vm.setFilterResults([{ countryName: 'foo' }]);
      await wrapper.setProps({
        filter: 'foo'
      });
      // Assert.
      expect(mock$emit).toHaveBeenCalledWith('suggestion', 'foo');
    });
    it('should not emit a suggestion after a tick if it doesn\'t have value', async () => {
      // Arrange.
      const wrapper = factory({
        setProps: {
          filter: 'foo'
        }
      });
      // Act.
      wrapper.vm.setFilterResults([{ countryName: 'foo' }]);
      await wrapper.setProps({
        filter: ''
      });
      // Assert.
      expect(mock$emit).not.toHaveBeenCalled();
    });
  });
  describe('computed', () => {
    const data = [{
      cases: 1,
      deaths: 2,
      recoveries: 3
    }, {
      cases: 1,
      deaths: 2,
      recoveries: 3
    }, {
      cases: 1,
      deaths: 2,
      recoveries: 3
    }];
    describe('totalCases', () => {
      it('should return the sum of the cases in the dataset', () => {
        // Arrange.
        const expected = 3;
        const wrapper = factory({
          propsData: { data }
        });
        // Act.
        // Assert.
        expect(wrapper.vm.totalCases).toStrictEqual(expected);
      });
      it('should return the sum of the deaths in the dataset', () => {
        // Arrange.
        const expected = 6;
        const wrapper = factory({
          propsData: { data }
        });
        // Act.
        // Assert.
        expect(wrapper.vm.totalDeaths).toStrictEqual(expected);
      });
      it('should return the sum of the recoveries in the dataset', () => {
        // Arrange.
        const expected = 9;
        const wrapper = factory({
          propsData: { data }
        });
        // Act.
        // Assert.
        expect(wrapper.vm.totalRecoveries).toStrictEqual(expected);
      });
    });
    it('should set the correct filter regex', async () => {
      // Arrange.
      const wrapper = factory({
        propsData: {
          filter: 'foo'
        }
      });
      // Act.
      // Assert.
      expect(wrapper.vm.filterRegEx).toEqual(/^foo/i);
      await wrapper.setProps({
        filter: 'bar'
      });
      expect(wrapper.vm.filterRegEx).toEqual(/^bar/i);
    });
  });
});
