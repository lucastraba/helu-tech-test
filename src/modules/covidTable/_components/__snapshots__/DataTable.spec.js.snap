// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Component conditional rendering should render the table provided it is loading and there are no errors 1`] = `
"<div class=\\"overflow-auto\\">
  <btable-stub sticky-header=\\"100vh\\" striped=\\"\\" show-empty=\\"\\" foot-clone=\\"\\" sort-direction=\\"desc\\" no-footer-sorting=\\"\\" fields=\\"[object Object],[object Object],[object Object],[object Object]\\" items=\\"\\" busy=\\"true\\" filter=\\"/^/i\\" filter-included-fields=\\"countryName\\"></btable-stub>
</div>"
`;

exports[`DataTable Component conditional rendering should render the table provided it is loading and there are no errors 2`] = `
"<div class=\\"overflow-auto\\">
  <div class=\\"error text-danger\\">
    foo
  </div>
</div>"
`;

exports[`DataTable Component conditional rendering should render the table provided there is data and there are no errors 1`] = `
"<div class=\\"overflow-auto\\">
  <btable-stub sticky-header=\\"100vh\\" striped=\\"\\" show-empty=\\"\\" foot-clone=\\"\\" sort-direction=\\"desc\\" no-footer-sorting=\\"\\" fields=\\"[object Object],[object Object],[object Object],[object Object]\\" items=\\"1\\" filter=\\"/^/i\\" filter-included-fields=\\"countryName\\"></btable-stub>
</div>"
`;
